### DISCUSSION TITLE
Copy Data from One Mimir Cluster to another

### DISCUSSION THREAD

**Discussion by JustNZ** [grafana/mimir#7563] [General]
Hello,

we currently have two Mimir Clusters and want to reduce the overall Infrastructure Size to only one.

So question now is, if it's possible to copy the bucket data from Cluster1 to the bucket of Cluster2 without any issues

**Comment by pstibrany** [comment]
Hello,
yes, copying blocks from long-term storage should cause no problems. You may want to disable compactors in source cluster (Cluster1) when you do that, to avoid blocks being changed (new blocks uploaded, old blocks marked for deletion) during the copy.

**Comment by JustNZ** [comment]
Hey,
alright cool - thanks for the fast answer

**Comment by JustNZ** [comment]
Hey @pstibrany ,

I did found an "issue" during copy of the blocks from Cluster1 to Cluster2:
1. inside the blocks s3 of Cluster1 is the same BockID as in the blocks s3 for Cluster2 - how to deal with that?
2. after moving a small amount of blocks from Cluster1 to Cluster2, the compactor on Mimir Cluster2 started to throw the following error:
`ts=2024-03-12T13:48:32.164077229Z caller=compactor.go:652 level=error component=compactor msg="failed to compact user blocks" user=test err="compaction: group 0@17241709254077376921-merge--1695859200000-1695945600000: compact blocks [/var/lib/mimir/data/data-compactor/compact/0@17241709254077376921-merge--1695859200000-1695945600000/01HBFDCC69FYA7DY6Q7N05TYXD /var/lib/mimir/data/data-compactor/compact/0@17241709254077376921-merge--1695859200000-1695945600000/01HBFCHQSCV2KY6HSC1PM6FRTM]: populate block: chunk iter: cannot populate chunk 42949673709 from block 01HBFCHQSCV2KY6HSC1PM6FRTM: checksum mismatch expected:3b12049, actual:f421572c"`

**Reply by pstibrany** [reply]
> inside the blocks s3 of Cluster1 is the same BockID as in the blocks s3 for Cluster2 - how to deal with that?

1. Are you saying that two clusters generated blocks with same ID? Can you share those IDs? I find that very strange given that we use cryptographic random number generator to produce those. Something seems off here.

2. This error message says that block `01HBFCHQSCV2KY6HSC1PM6FRTM` has corrupted chunks. Compactor can't work with it, and block should be marked for non-compaction. (We have `markblocks` tool for that in Mimir tools repository)

**Reply by JustNZ** [reply]
Hey @pstibrany ,
here is for example a few of blocks with the same IDs on both clusters
```
01H8G30YH46MCFWK10PMXJN2KW
01H8K2R3AZT7N9JSJFF2NDGEA9
01H8N8E7SCB3PBBD6PERMJ7PEJ
01H8QTZPEXGG8TEQFBKY0BPQMF
01H8TDCKH5WM2D56KSGNATE9RT
01H8X1ZFE5BED94C7JX88ZF9K4
01H8ZN7G70Z28DDCXTPF7H8DJ4
01H926WF74G7YFWZ1394S7616G
01H94S30Z0VZQP3GWQX2X2X524
01H97970YAQ4EKCZ6QS1QB4Q27
01H99YVZ0ACNDYXF6FMWM91V4S
01H9CFBYCS0XHYNW10X1HC8NSQ
01H9F454S0KD1VAYW53225D60D
01H9HKGZ8PDWSEJYMPRZBG65YN
```

**Reply by pstibrany** [reply]
These blocks have following timestamps and random parts. I still find it hard to believe that two clusters would use exact same timestamps and random bytes in generated block IDs. 🤔 

```
ULID                        TIMESTAMP                 RANDOM
01H8G30YH46MCFWK10PMXJN2KW  2023-08-23T02:19:22.532Z  3518fe4c20b53b2a8a7c
01H8K2R3AZT7N9JSJFF2NDGEA9  2023-08-24T06:12:15.839Z  d1ea99664f78aad83949
01H8N8E7SCB3PBBD6PERMJ7PEJ  2023-08-25T02:30:13.036Z  58ecb5b4d6762923d9d2
01H8QTZPEXGG8TEQFBKY0BPQMF  2023-08-26T02:32:48.349Z  8411a75deb9f80bb5e8f
01H8TDCKH5WM2D56KSGNATE9RT  2023-08-27T02:32:54.565Z  e504d29a798555a7271a
01H8X1ZFE5BED94C7JX88ZF9K4  2023-08-28T03:11:13.349Z  5b9a9230f2ea11f7a664
01H8ZN7G70Z28DDCXTPF7H8DJ4  2023-08-29T03:26:08.096Z  f890d6b3bab3cf143644
01H926WF74G7YFWZ1394S7616G  2023-08-30T03:13:09.86Z   81fcfe7c2349327304d0
01H94S30Z0VZQP3GWQX2X2X524  2023-08-31T03:09:47.872Z  dfef61c397e8ba2e9444
01H97970YAQ4EKCZ6QS1QB4Q27  2023-09-01T02:30:05.002Z  b91d367cd7c86eb25c47
01H99YVZ0ACNDYXF6FMWM91V4S  2023-09-02T03:27:00.106Z  655beebccfa72890ec99
01H9CFBYCS0XHYNW10X1HC8NSQ  2023-09-03T02:53:49.849Z  0763eaf020e862c45737
01H9F454S0KD1VAYW53225D60D  2023-09-04T03:35:35.968Z  9b43b57b85188456980d
01H9HKGZ8PDWSEJYMPRZBG65YN  2023-09-05T02:42:41.046Z  6f32e97a96c7d70317d5
```

**Reply by JustNZ** [reply]
Stupid question, but if I move an block with the same ID which exists in both buckets from bucket A to B it will override the content - correct?

here is btw. the "prove" that the same object exists in two different buckets
![image](https://github.com/grafana/mimir/assets/51332147/0489ed7a-c2fb-4972-9d2b-5558ff90847b)

**Reply by pstibrany** [reply]
> Stupid question, but if I move an block with the same ID which exists in both buckets from bucket A to B it will override the content - correct?

Yes. I wonder if blocks are exactly the same as well? (Eg. could you run some hash on files? Perhaps `mc` can report md5?)

> 
> here is btw. the "prove" that the same object exists in two different buckets ![image](https://private-user-images.githubusercontent.com/51332147/312396478-0489ed7a-c2fb-4972-9d2b-5558ff90847b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzMjcyNjAsIm5iZiI6MTcxMDMyNjk2MCwicGF0aCI6Ii81MTMzMjE0Ny8zMTIzOTY0NzgtMDQ4OWVkN2EtYzJmYi00OTcyLTlkMmItNTU1OGZmOTA4NDdiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzEzVDEwNDkyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg0NjBlZTgzYjBjZmM2ZDQ4MjM1YzM2MjA5MjNhMTA1ZmY3N2FkZGVkOTMyZTU4NWM4ZjcwMTQ3M2JjMTJiOWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZNN0hytdZGD6zI-U3zBPFJriq0MFRtjZXX2eLsJ2_04)

I trust that you see what you see :) but I don't think it's possibly generated by two separate clusters. Perhaps objects are somehow copied or duplicated between the buckets?

**Reply by JustNZ** [reply]
So on my screenshot we can see the following:
- Left Side = Mimir Prod Cluster with all the content of our prod servers
- Right Side = A copy to the new S3 bucket from the Mimir NonProd Cluster with content of all nonprod servers

So I would assume the ID of the block and the chunks itself are the same - but the data in there has to be different at least.

In the end if we can't keep our data and have to start from the beginning again at the new bucket - we will do that

Chunk Screenshot from one block on both s3
![image](https://github.com/grafana/mimir/assets/51332147/9a4e5182-c158-47f4-94ac-c340313ebc49)

**Reply by pstibrany** [reply]
IDs alone indicate it's the same block to me. Same size confirms that. I don't think you need to start from scratch (although it would be good to understand how blocks got duplicated).

To be sure that blocks are correct, you could download them locally and use `tsdb-index-health` tool from Mimir repository, with `-check-chunks` option to check chunks as well. (This will print some details about the block, and could also confirm that these blocks are exactly the same).

**Reply by JustNZ** [reply]
I did the `tsdb-index-health` on the block which is in both buckets and it does indeed look the same. But that confuses me a lot:

![image](https://github.com/grafana/mimir/assets/51332147/c8c6cf4b-85b3-4dd0-9acc-a0f6f4388b0c)

**Reply by JustNZ** [reply]
Okay I think I now got a plan how to deal with that situation. Thanks a lot so far for your help.

Now my last question - if I copy the blocks from one Mimir Cluster s3 to the other one. How can I tell the Target Cluster to "use" these new blocks. Because as I said I already copied yesterday a lot of data but still can't query that

**Reply by JustNZ** [reply]
Okay now I can also see the "old" data - don't ask me how and why, but it works :D 

Thanks a lot @pstibrany for all your effort to assist me on that - very appreciate that

**Reply by pstibrany** [reply]
> Okay now I can also see the "old" data - don't ask me how and why, but it works :D

There are some delays (compactor needs to find the blocks, and include them into bucket index, then store-gateways and queriers need to reload bucket index), but after a while, blocks should be queryable.
